### ISSUE TITLE
Filter Map on Model

### ISSUE THREAD

**Issue by CarlosLeffa** [abbasudo/laravel-purity#6]
Hello. First congratulations on the excellent package.

I am using this package in an api. But I use Resources for formatting and in that case some column/relationship names are overwritten.
```
...
public function toArray($request)
    {
        return [
            'id' => $this->id,
            'name'   => $this->title,
            'description'   => $this->description,
            'date'  => $this->created_at,
            'last_update'  => $this->updated_at,
        ];
    }
...
```

My idea is to allow the API client to use the filter with the names of the columns/relationships that are in its Resource and define which columns/relationships will be allowed in the filters.

To solve this, I made a small change that I wanted to share.

- src/Exceptions/NoFilterMap.php
> Create new exception class
- src/Traits/Filterable.php
> I added a new property named $filtersMap and a public function to return the property
- src/Filters/Resolve.php
> Added a private relationFiltersMap() function to resolve the column/relationship name
```
...
    private function relationFiltersMap(Builder $query, string $field): string
    {
        $filtersMap = $query->getModel()->getFiltersMap();
        if ($filtersMap) {
            if (array_key_exists($field, $filtersMap)) {
                $field = $filtersMap[$field];
            } else{
                throw NoFilterMap::create($field, $query->getModel()::class);
            }
        }
        return $field;
    }
...
```

> In the applyFilterStrategy() method I changed the value of $callback to a method that calls relationFiltersMap() to adjust the $field.
> In relation() method I changed the value of $field to a method that calls relationFiltersMap().

- In the Models the following Key/Value array can be defined. If this property is defined in the Model, the filter will be possible only in predefined columns in the array.
```
...
class Post extends Model
{
    use HasFactory;
    use Filterable, Sortable;

    protected array $filtersMap = 
    [
        "id" => "id",
        "name" => "title",
        "date" => "created_at",
        "last_update" => "updated_at",
        "Interactions" => "comments",  // hasMany relationship 
    ];
...
```

**If you want to incorporate this small implementation in your package, follow the attachment.**
[FiltersMap.zip](https://github.com/abbasudo/laravel-purity/files/11544393/FiltersMap.zip)

**Comment by abbasudo** [comment]
Thank you for your feedback and for sharing your implementation with me. I appreciate your contribution to the package.

allow columns were in this package roadmap and I got inspired by your implementation. I will definitely add it, but there are some points that I need to consider and modify in your code.

Firstly, I carefully implemented this code with simplicity of usage in mind but also made sure to provide ways for advanced usage. The `filter()` method can work with or without any parameters, and the code can work with or without `$filters` in the model. Your implementation forces the model to have `$filtersMap`. changing this behavior is not that hard ill do it easily.

Secondly, the package supports nested filters, so you can filter posts by columns of the user model, even if the post model is related to the comment model, which is related to the user model. In your implementation of the filter map, we have to specify the user's columns in the post model, which is not ideal. 

So, I need help to modify your code to make sure that nested filters are still possible without having to specify the columns in the wrong model. Thank you again for your contribution, and I will keep you updated on the progress.

**Comment by abbasudo** [comment]
this feature is implemented in #17. now you can rename and restrict allowed filters on the model or controller.
[allowed-fields](https://github.com/abbasudo/laravel-purity#allowed-fields)
[rename-fields](https://github.com/abbasudo/laravel-purity#rename-fields)
